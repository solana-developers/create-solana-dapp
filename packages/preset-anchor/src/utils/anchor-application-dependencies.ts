import { addDependenciesToPackageJson, Tree } from '@nx/devkit'
import { packageVersion } from '@solana-developers/preset-common'

export function anchorApplicationDependencies(tree: Tree) {
  return addDependenciesToPackageJson(
    tree,
    {
      '@coral-xyz/anchor': packageVersion['@coral-xyz'].anchor,
    },
    {},
  )
}
