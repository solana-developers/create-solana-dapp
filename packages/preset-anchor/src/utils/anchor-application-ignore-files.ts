import { Tree } from '@nx/devkit'
import { updateIgnoreEntries } from '@solana-developers/preset-common'

export function anchorApplicationIgnoreFiles(tree: Tree, projectRoot: string) {
  const ignoreFiles = {
    git: [
      '.anchor',
      `${projectRoot}/target/deploy`,
      `${projectRoot}/target/release`,
      `${projectRoot}/target/sbf-solana-solana`,
      `${projectRoot}/target/.rustc_info.json`,
      `!${projectRoot}/target/idl/*.json`,
      `!${projectRoot}/target/types/*.ts`,
      'node_modules',
      'test-ledger',
      '.yarn',
    ],
    prettier: ['.anchor', 'target', 'node_modules', 'dist', 'tmp', 'build', 'test-ledger'],
  }

  updateIgnoreEntries(tree, '.gitignore', ignoreFiles.git)
  updateIgnoreEntries(tree, '.prettierignore', ignoreFiles.prettier)
}
