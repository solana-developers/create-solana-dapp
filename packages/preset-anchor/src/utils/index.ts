export * from './anchor-application-ignore-files'
export * from './anchor-application-dependencies'
export * from './anchor-application-normalize-schema'
