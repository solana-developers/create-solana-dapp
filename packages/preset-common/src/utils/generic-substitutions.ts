import { names } from '@nx/devkit'

export interface GenericSubstitutions {
  name: string
  npmScope: string
  anchor: string
  anchorName: string
  preset?: 'next' | 'react'
}
export function genericSubstitutions({ anchor, anchorName, name, npmScope, preset }: GenericSubstitutions) {
  return {
    ...names(name),
    anchor: names(anchor),
    anchorName: names(anchorName),
    npmScope,
    currentFullYear: new Date().getFullYear(),
    preset: preset ?? 'react',
  }
}
