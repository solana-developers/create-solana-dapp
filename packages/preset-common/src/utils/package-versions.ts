export const packageVersion = {
  '@coral-xyz': {
    anchor: '^0.29.0',
  },
  '@tailwindcss': {
    typography: '0.5.10',
  },
  '@solana': {
    'spl-token': '0.3.9',
    'wallet-adapter-base': '^0.9.23',
    'wallet-adapter-react': '^0.15.35',
    'wallet-adapter-react-ui': '^0.9.34',
    'wallet-adapter-solflare': '^0.6.28',
    'wallet-adapter-unsafe-burner': '^0.1.7',
    'web3.js': '1.87.2',
  },
  '@swc-node': {
    core: '1.10.6',
    register: '1.6.8',
  },
  '@tabler': {
    'icons-react': '2.42.0',
  },
  '@tanstack': {
    'react-query': '5.8.4',
  },
  'react-hot-toast': '2.4.1',
  bs58: '5.0.0',
  buffer: '6.0.3',
  daisyui: '3.9.3',
  encoding: '0.1.13',
  jotai: '2.5.1',
}
