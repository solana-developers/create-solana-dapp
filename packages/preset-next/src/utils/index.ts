export * from './next-application-substitutions'
export * from './generate-next-application'
export * from './normalize-next-application-schema'
