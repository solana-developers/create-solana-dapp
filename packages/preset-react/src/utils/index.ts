export * from './application-tailwind-config'
export * from './generate-react-application'
export * from './get-ui-dependencies'
export * from './normalize-react-application-schema'
export * from './react-application-dependencies'
export * from './react-application-run-scripts'
export * from './react-application-substitutions'
export * from './wallet-adapter-dependencies'
