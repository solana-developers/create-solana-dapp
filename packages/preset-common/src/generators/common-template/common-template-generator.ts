import { generateFiles, Tree } from '@nx/devkit'
import * as path from 'path'
import { genericSubstitutions } from '../../utils/generic-substitutions'
import { CommonTemplateSchema } from './common-template-schema'

export async function commonTemplateGenerator(tree: Tree, options: CommonTemplateSchema, templatePath: string = '') {
  const substitutions = genericSubstitutions({
    anchor: options.anchor,
    anchorName: options.anchorName,
    licenseAuthor: options.licenseAuthor,
    name: options.webName,
    npmScope: options.npmScope,
    preset: options.preset,
  })
  generateFiles(tree, path.join(__dirname, 'files', options.template, templatePath), options.directory, {
    ...options,
    ...substitutions,
    fileNameUnderscore: substitutions.fileName.replace(/-/g, '_'),
  })
}

export default commonTemplateGenerator
