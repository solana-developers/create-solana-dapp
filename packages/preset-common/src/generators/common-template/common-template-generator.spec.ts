import { Tree } from '@nx/devkit'
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing'
import { getRecursiveFileContents } from '../../utils/get-recursive-file-contents'
import { commonTemplateGenerator } from './common-template-generator'
import { CommonTemplateSchema } from './common-template-schema'

describe('common-template generator', () => {
  let tree: Tree
  const options: CommonTemplateSchema = {
    name: 'test',
    template: 'license',
    directory: 'test',
    anchor: 'none',
    anchorName: 'anchor',
    npmScope: 'test',
    webName: 'test',
  }
  const templates: CommonTemplateSchema['template'][] = ['license', 'readme']
  const anchor: CommonTemplateSchema['anchor'][] = ['none', 'counter', 'hello-world']

  const matrix = templates.map((template) => anchor.map((anchor) => ({ template, anchor }))).flat()
  process.env['USER'] = 'test'

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace()
  })

  it.each(matrix)('should generate files for %s template with %s anchor', async ({ template, anchor }) => {
    await commonTemplateGenerator(tree, { ...options, template, anchor })

    const contents = getRecursiveFileContents(tree, options.directory)
    expect(contents).toMatchSnapshot()
  })
})
