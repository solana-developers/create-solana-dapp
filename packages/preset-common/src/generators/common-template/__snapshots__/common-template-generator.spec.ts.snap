// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`common-template generator should generate files for { template: 'license', anchor: 'counter' } template with %s anchor 1`] = `
{
  "LICENSE": {
    "content": [
      "MIT License",
      "Copyright (c) 2023 test",
      "Permission is hereby granted, free of charge, to any person obtaining a copy",
      "of this software and associated documentation files (the "Software"), to deal",
      "in the Software without restriction, including without limitation the rights",
      "to use, copy, modify, merge, publish, distribute, sublicense, and/or sell",
      "copies of the Software, and to permit persons to whom the Software is",
      "furnished to do so, subject to the following conditions:",
      "The above copyright notice and this permission notice shall be included in all",
      "copies or substantial portions of the Software.",
      "THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR",
      "IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,",
      "FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE",
      "AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER",
      "LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,",
      "OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE",
      "SOFTWARE.",
    ],
    "isBinary": false,
    "path": "test/LICENSE",
  },
}
`;

exports[`common-template generator should generate files for { template: 'license', anchor: 'hello-world' } template with %s anchor 1`] = `
{
  "LICENSE": {
    "content": [
      "MIT License",
      "Copyright (c) 2023 test",
      "Permission is hereby granted, free of charge, to any person obtaining a copy",
      "of this software and associated documentation files (the "Software"), to deal",
      "in the Software without restriction, including without limitation the rights",
      "to use, copy, modify, merge, publish, distribute, sublicense, and/or sell",
      "copies of the Software, and to permit persons to whom the Software is",
      "furnished to do so, subject to the following conditions:",
      "The above copyright notice and this permission notice shall be included in all",
      "copies or substantial portions of the Software.",
      "THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR",
      "IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,",
      "FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE",
      "AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER",
      "LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,",
      "OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE",
      "SOFTWARE.",
    ],
    "isBinary": false,
    "path": "test/LICENSE",
  },
}
`;

exports[`common-template generator should generate files for { template: 'license', anchor: 'none' } template with %s anchor 1`] = `
{
  "LICENSE": {
    "content": [
      "MIT License",
      "Copyright (c) 2023 test",
      "Permission is hereby granted, free of charge, to any person obtaining a copy",
      "of this software and associated documentation files (the "Software"), to deal",
      "in the Software without restriction, including without limitation the rights",
      "to use, copy, modify, merge, publish, distribute, sublicense, and/or sell",
      "copies of the Software, and to permit persons to whom the Software is",
      "furnished to do so, subject to the following conditions:",
      "The above copyright notice and this permission notice shall be included in all",
      "copies or substantial portions of the Software.",
      "THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR",
      "IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,",
      "FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE",
      "AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER",
      "LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,",
      "OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE",
      "SOFTWARE.",
    ],
    "isBinary": false,
    "path": "test/LICENSE",
  },
}
`;

exports[`common-template generator should generate files for { template: 'readme', anchor: 'counter' } template with %s anchor 1`] = `
{
  "README.md": {
    "content": [
      "# test",
      "This project is generated with the [create-solana-dapp](https://github.com/solana-developers/create-solana-dapp) generator.",
      "## Getting Started",
      "### Prerequisites",
      "- Node v18.18.0 or higher",
      "- Rust v1.70.0 or higher",
      "- Anchor CLI 0.29.0 or higher",
      "- Solana CLI 1.17.0 or higher",
      "### Installation",
      "#### Clone the repo",
      "\`\`\`shell",
      "git clone <repo-url>",
      "cd <repo-name>",
      "\`\`\`",
      "#### Install Dependencies",
      "\`\`\`shell",
      "yarn install",
      "\`\`\`",
      "#### Start the web app",
      "\`\`\`",
      "yarn dev",
      "\`\`\`",
      "## Apps",
      "### anchor",
      "This is a Solana program written in Rust using the Anchor framework.",
      "#### Commands",
      "You can use any normal anchor commands. Either move to the \`anchor\` directory and run the \`anchor\` command or prefix the command with \`yarn\`, eg: \`yarn anchor\`.",
      "#### Sync the program id:",
      "Running this command will create a new keypair in the \`anchor/target/deploy\` directory and save the address to the Anchor config file and update the \`declare_id!\` macro in the \`./src/lib.rs\` file of the program.",
      "You will manually need to update the constant in \`anchor/lib/counter-exports.ts\` to match the new program id.",
      "\`\`\`shell",
      "yarn anchor keys sync",
      "\`\`\`",
      "#### Build the program:",
      "\`\`\`shell",
      "yarn anchor-build",
      "\`\`\`",
      "#### Start the test validator with the program deployed:",
      "\`\`\`shell",
      "yarn anchor-localnet",
      "\`\`\`",
      "#### Run the tests",
      "\`\`\`shell",
      "yarn anchor-test",
      "\`\`\`",
      "#### Deploy to Devnet",
      "\`\`\`shell",
      "yarn anchor deploy --provider.cluster devnet",
      "\`\`\`",
      "### web",
      "This is a React app that uses the Anchor generated client to interact with the Solana program.",
      "#### Commands",
      "Start the web app",
      "\`\`\`shell",
      "yarn dev",
      "\`\`\`",
      "Build the web app",
      "\`\`\`shell",
      "yarn build",
      "\`\`\`",
    ],
    "isBinary": false,
    "path": "test/README.md",
  },
}
`;

exports[`common-template generator should generate files for { template: 'readme', anchor: 'hello-world' } template with %s anchor 1`] = `
{
  "README.md": {
    "content": [
      "# test",
      "This project is generated with the [create-solana-dapp](https://github.com/solana-developers/create-solana-dapp) generator.",
      "## Getting Started",
      "### Prerequisites",
      "- Node v18.18.0 or higher",
      "- Rust v1.70.0 or higher",
      "- Anchor CLI 0.29.0 or higher",
      "- Solana CLI 1.17.0 or higher",
      "### Installation",
      "#### Clone the repo",
      "\`\`\`shell",
      "git clone <repo-url>",
      "cd <repo-name>",
      "\`\`\`",
      "#### Install Dependencies",
      "\`\`\`shell",
      "yarn install",
      "\`\`\`",
      "#### Start the web app",
      "\`\`\`",
      "yarn dev",
      "\`\`\`",
      "## Apps",
      "### anchor",
      "This is a Solana program written in Rust using the Anchor framework.",
      "#### Commands",
      "You can use any normal anchor commands. Either move to the \`anchor\` directory and run the \`anchor\` command or prefix the command with \`yarn\`, eg: \`yarn anchor\`.",
      "#### Sync the program id:",
      "Running this command will create a new keypair in the \`anchor/target/deploy\` directory and save the address to the Anchor config file and update the \`declare_id!\` macro in the \`./src/lib.rs\` file of the program.",
      "You will manually need to update the constant in \`anchor/lib/hello-world-exports.ts\` to match the new program id.",
      "\`\`\`shell",
      "yarn anchor keys sync",
      "\`\`\`",
      "#### Build the program:",
      "\`\`\`shell",
      "yarn anchor-build",
      "\`\`\`",
      "#### Start the test validator with the program deployed:",
      "\`\`\`shell",
      "yarn anchor-localnet",
      "\`\`\`",
      "#### Run the tests",
      "\`\`\`shell",
      "yarn anchor-test",
      "\`\`\`",
      "#### Deploy to Devnet",
      "\`\`\`shell",
      "yarn anchor deploy --provider.cluster devnet",
      "\`\`\`",
      "### web",
      "This is a React app that uses the Anchor generated client to interact with the Solana program.",
      "#### Commands",
      "Start the web app",
      "\`\`\`shell",
      "yarn dev",
      "\`\`\`",
      "Build the web app",
      "\`\`\`shell",
      "yarn build",
      "\`\`\`",
    ],
    "isBinary": false,
    "path": "test/README.md",
  },
}
`;

exports[`common-template generator should generate files for { template: 'readme', anchor: 'none' } template with %s anchor 1`] = `
{
  "README.md": {
    "content": [
      "# test",
      "This project is generated with the [create-solana-dapp](https://github.com/solana-developers/create-solana-dapp) generator.",
      "## Getting Started",
      "### Prerequisites",
      "- Node v18.18.0 or higher",
      "### Installation",
      "#### Clone the repo",
      "\`\`\`shell",
      "git clone <repo-url>",
      "cd <repo-name>",
      "\`\`\`",
      "#### Install Dependencies",
      "\`\`\`shell",
      "yarn install",
      "\`\`\`",
      "#### Start the web app",
      "\`\`\`",
      "yarn dev",
      "\`\`\`",
      "## Apps",
      "### web",
      "This is a React app.",
      "#### Commands",
      "Start the web app",
      "\`\`\`shell",
      "yarn dev",
      "\`\`\`",
      "Build the web app",
      "\`\`\`shell",
      "yarn build",
      "\`\`\`",
    ],
    "isBinary": false,
    "path": "test/README.md",
  },
}
`;
